### PR TITLE
Refactor ``find_default_config_files``

### PR DESCRIPTION
**Pull request by DanielNoord**
- [x] Add yourself to CONTRIBUTORS if you are a new contributor.
- [x] Write a good description on what the PR does.

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :hammer: Refactoring   |

## Description

I have integrated the tests from #6056.

We have a similar issue here: `find_default_config_files` is in `__all__` but it is such an internal function that I think we should be able to change it.
I can also add `str()` calls to all `yields`.

### CODE REVIEW

File: pylint/lint/pylinter.py [hunk 1]
```diff
@@ -612,9 +612,10 @@ def __init__(
         self._by_id_managed_msgs: List[ManagedMessage] = []
 
         reporters.ReportsHandlerMixIn.__init__(self)
+        config_file = next(config.find_default_config_files(), None)
         super().__init__(
             usage=__doc__,
-            config_file=pylintrc or next(config.find_default_config_files(), None),
+            config_file=pylintrc or str(config_file) if config_file else None,
```
Comment (DanielNoord) [inline pylint/lint/pylinter.py:618]: We will remove this in #6052 so this is only temporary.

File: pylint/config/find_default_config_files.py [hunk 2]
```diff
@@ -29,54 +33,52 @@ def _toml_has_config(path):
     return True
 
 
-def _cfg_has_config(path):
+def _cfg_has_config(path: Path) -> bool:
     parser = configparser.ConfigParser()
     parser.read(path, encoding="utf-8")
     return any(section.startswith("pylint.") for section in parser.sections())
```
Comment (DanielNoord) [inline pylint/config/find_default_config_files.py:39]: @Pierre-Sassoulas I have not copied the handling of malformed `setup.cfg`. If it is malformed we should raise a warning like we do for `.toml` so we should probably handle that in another place where we have access to the `linter` object.

Comment (Pierre-Sassoulas) [reply pylint/config/find_default_config_files.py:39]: It's inside a function that check if the setup.cfg contains a pylint configuration, not handling it means a malformed / unused setup.cfg would crash pylint. Could we discuss and merge the other MR first ?

File: pylint/config/find_default_config_files.py [hunk 3]
```diff
@@ -29,54 +33,52 @@ def _toml_has_config(path):
...
     else:
-        user_home = os.path.expanduser("~")
-        if user_home not in ("~", "/root"):
-            home_rc = os.path.join(user_home, ".pylintrc")
-            if os.path.isfile(home_rc):
-                yield home_rc
-            home_rc = os.path.join(user_home, ".config", "pylintrc")
-            if os.path.isfile(home_rc):
-                yield home_rc
+        user_home = Path("~").expanduser()
```
Comment (Pierre-Sassoulas) [inline pylint/config/find_default_config_files.py:66]: Unrelated to the current refactor, but this part is suspicious, we have a USER_HOME in ``pylint.constant``.

**Comment by DanielNoord** [comment]
> ## Pull Request Test Coverage Report for [Build 2070480600](https://coveralls.io/builds/47868665)
> * **47** of **52**   **(90.38%)**  changed or added relevant lines in **4** files are covered.
> * **1** unchanged line in **1** file lost coverage.
> * Overall coverage increased (+**0.0005%**) to **94.311%**
> 
> Changes Missing Coverage	Covered Lines	Changed/Added Lines	%
> [pylint/testutils/utils.py](https://coveralls.io/builds/47868665/source?filename=pylint%2Ftestutils%2Futils.py#L34)	16	17	94.12%
> [pylint/config/find_default_config_files.py](https://coveralls.io/builds/47868665/source?filename=pylint%2Fconfig%2Ffind_default_config_files.py#L62)	29	33	87.88%
> Files with Coverage Reduction	New Missed Lines	%
> [pylint/config/find_default_config_files.py](https://coveralls.io/builds/47868665/source?filename=pylint%2Fconfig%2Ffind_default_config_files.py#L45)	1	88.52%
> Totals	[Coverage Status](https://coveralls.io/builds/47868665)
> Change from base [Build 2070459980](https://coveralls.io/builds/47868417):	0.0005%
> Covered Lines:	15483
> Relevant Lines:	16417
> ##### 💛 - [Coveralls](https://coveralls.io)

https://coveralls.io/builds/47868417/source?filename=pylint%2Fconfig%2Ffind_default_config_files.py

Doesn't decrease. Just badly covered currently.

File: pylint/config/find_default_config_files.py [hunk 4]
```diff
@@ -34,43 +37,41 @@ def _cfg_has_config(path: Union[Path, str]) -> bool:
...
+        if str(user_home) not in ("~", "/root"):
```
Comment (Pierre-Sassoulas) [inline pylint/config/find_default_config_files.py:65]: This look like it could also be refactored but let's focus on os.path => pathlib for now.

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]
LGTM !

**Review by Pierre-Sassoulas — APPROVED** [review]

**Pull request merged**